sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/internal/ResourceDetector.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.sdk.autoconfigure.spi.internal;
+
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.sdk.autoconfigure.spi.ConfigProperties;
+import io.opentelemetry.sdk.autoconfigure.spi.Ordered;
+import java.util.Optional;
+import java.util.function.Function;
+
+public interface ResourceDetector<D> extends Ordered {
```

Comment:
I really don't see the current ResourceProvider design as broken, and can't tell how this would alleviate confusion. 

Here's a summary of the toolkit for managing resources today:
- ResourceProvider SPI allows you provide your own resource, which is merged with resource provided by other providers
- The environment variables `OTEL_SERVICE_NAME` and `OTEL_RESOURCE_ATTRIBUTES` contribute to the environment resource provider, which runs last. These environment variables can be customized by `AutoConfigurationCustomizer#addPropertiesSupplier`, `AutoConfigurationCustomizer#addPropertiesCustomizer`.
- ResourceProvider can optionally be ordered, so you can indicate priority relative to others
- ResourceProvider can optionally be conditional, so you can decide whether or not to provide a resource based on the resource computed up till that point in time
- All resource providers are enabled by default, and you can set a deny list with `otel.java.disabled.resource.providers`
- All resource providers are enabled by default, but you can optionally change the semantic to an allow list by setting `otel.java.enabled.resource.providers`
- After all resource providers have been applied, you can optionally filter out individual attribute keys with `otel.experimental.resource.disabled.keys`
- After all resource providers have been applied and optional filter applied, you can customize the resource further with `AutoConfigurationCustomizer#addResourceCustomizer`
- After the resource has been completely resolved and applied to `Sdk{Signal}ProviderBuilder`, you can further customize it with `AutoConfigurationCustomizer#add{Signal}ProviderCustomizer` via `Sdk{Signal}ProviderBuilder`'s `setResource` and `addResource` methods.

That's quite a long and comprehensive list, with interactions which are hard to reason through for a casual user. I'm very reluctant to add additional complexity unless there is a very strong reason.

Reply:
From what I've learned, there's one idiomatic way how to write resource providers, which is easy to get wrong. 
Do the idea is to take complexity away, not to add it.
At the same time, this is not a topic that is blocking any of my current tasks.